Adding Two Factor Auth Feature

tnylea opened the pull request:
This PR adds **Two Factor Authentication** to this starter kit.

On the user settings page, a new tab called **Two-Factor Auth** will be added (we can rename this tab if desired).

![CleanShot 2025-04-17 at 15 24 07@2x](https://github.com/user-attachments/assets/256a4a8e-f2dd-4b5e-93ab-dd1edb3c3619)

When the user clicks that tab, they will see the following page which notifies them that 2FA is currently **Disabled**.

![CleanShot 2025-04-17 at 15 24 21@2x](https://github.com/user-attachments/assets/b7d4cc90-2f08-493e-9b1e-44ffa22bdc7c)


Clicking the **Enable** button will open a modal with a QR code and an alpha-numeric code. The user can enter this into their 2FA application (Google Authenticator).

![CleanShot 2025-04-17 at 15 13 14@2x](https://github.com/user-attachments/assets/ae4073e2-5a55-4edf-a41f-7d5e588ae0f7)

Next, after clicking Continue, they’ll be prompted to enter the authentication code from their authenticator app.

![CleanShot 2025-04-17 at 15 13 21@2x](https://github.com/user-attachments/assets/e25e348e-30ec-4ce8-bd10-931d35be6fba)

If the user enters the correct code the modal will close and they will see the following screen.

![CleanShot 2025-04-17 at 15 12 32@2x](https://github.com/user-attachments/assets/b77a859b-1cf9-438d-8c05-b8f4fe3a14ce)

This shows the user that 2FA is now **Enabled** and provides them with Recovery codes. Now, when the user logs in to their account, they will see an additional authentication page where they are prompted for their 2FA code.

![CleanShot 2025-04-17 at 15 14 03@2x](https://github.com/user-attachments/assets/d3dbc219-b8ca-43c7-90e5-e17a009e47e9)

The user may also click the **Use Recovery Codes** button, and they will be presented with an input to enter one of their recovery codes. A recovery code can only be used once. After the user enters the recovery code, it will no longer be valid. If the user adds all the codes to the input, the logic will strip out the first code and verify it.

The 2FA challenge page utilizes the `auth` layout, so it will look good on any layout the developer chooses.

![CleanShot 2025-04-17 at 15 14 30@2x](https://github.com/user-attachments/assets/78c1f771-a578-4fed-837f-56fdd15ceabb)

![CleanShot 2025-04-17 at 15 22 47@2x](https://github.com/user-attachments/assets/26c6befb-15a8-45ed-bf85-4c88adc9ab71)

**Other Starter Kit 2FA Feature PR's**

React: [github.com/laravel/react-starter-kit/pull/101](https://github.com/laravel/react-starter-kit/pull/101)
Vue: [github.com/laravel/vue-starter-kit/pull/120](https://github.com/laravel/vue-starter-kit/pull/120)

Jubeki commented on app/Actions/TwoFactorAuth/GenerateQrCodeAndSecretKey.php:
```diff
@@ -0,0 +1,50 @@
+        $writer = new Writer(
+            new ImageRenderer(
+                new RendererStyle(800),
+                new ImagickImageBackEnd
```
Not sure if the Imagick backend should be used. Because it is not a requirement for laravel itself.

Maybe use SvgImageBackEnd or the GDLib Renderer?
https://github.com/Bacon/BaconQrCode/blob/main/README.md#available-image-renderer-back-ends

tnylea replied:
For sure! Thanks for the quick comment 👍 

That's a good call. I'll try out the SVGImageBackEnd and see how that works.

tnylea replied:
Thanks @Jubeki 👏

I've tested and implemented using the `SvgImageBackEnd()`, which is better than using the Imagick base64 PNG. Appreciate the input!

lucasjose501 commented on database/migrations/0001_01_01_000003_add_two_factor_auth_columns.php:
```diff
@@ -0,0 +1,34 @@
+     */
+    public function up(): void
+    {
+        Schema::table('users', function (Blueprint $table) {
```
This could be simplified by using the after callback:

            $table->after('password', function (Blueprint $table) {
                $table->text('two_factor_secret')->nullable();
                $table->text('two_factor_recovery_codes')->nullable();
                $table->timestamp('two_factor_confirmed_at')->nullable();
            });

tnylea replied:
Nice! I didn't know about that. Just updated 👏

Thanks for the input.

lucasjose501 commented:
I love this implementation, I did something similar in my project, and while I was thinking about security, I think there are two things to consider:

1. Make the database secret and recovery codes encrypted, someone with access to the database would be able to bypass the login with that information.
2. Ask for password confirmation before enabling/disabling, and showing recovery codes.

I'm not an expert in security, this is just something that crossed my mind.

tnylea commented:
> I love this implementation, I did something similar in my project, and while I was thinking about security, I think there are two things to consider:
> 
> 1. Make the database secret and recovery codes encrypted, someone with access to the database would be able to bypass the login with that information.
> 2. Ask for password confirmation before enabling/disabling, and showing recovery codes.
> 
> I'm not an expert in security, this is just something that crossed my mind.

Thanks @lucasjose501, the secrets and recovery codes are encrypted 👍

crynobone commented on app/Actions/TwoFactorAuth/VerifyTwoFactorCode.php:
```diff
@@ -0,0 +1,21 @@
+     * @param  string  $code The code to verify
+     * @return bool
+     */
+    public function __invoke(string $secret, string $code): bool
+    {
```
```suggestion
    public function __invoke(
        #[\SensitiveParameter] string $secret, 
        #[\SensitiveParameter] string $code
    ): bool {
```

tnylea replied:
Great call! I wasn't aware of that attribute. Thanks!

crynobone commented on app/Actions/TwoFactorAuth/ProcessRecoveryCode.php:
```diff
@@ -0,0 +1,34 @@
+     * @param  string  $submittedCode The code submitted by the user
+     * @return array|false Returns the updated array of recovery codes if valid, or false if invalid
+     */
+    public function __invoke(array $recoveryCodes, string $submittedCode)
+    {
```
```suggestion
    public function __invoke(
        #[\SensitiveParameter] array $recoveryCodes, 
        #[\SensitiveParameter] string $submittedCode
    ) {
```

crynobone reviewed (commented):
I would be useful if we utilise `#[\SensitiveParameter]` attribute: https://php.watch/versions/8.2/backtrace-parameter-redaction#SensitiveParameter

valorin commented on app/Actions/TwoFactorAuth/ProcessRecoveryCode.php:
```diff
@@ -0,0 +1,34 @@
+        
+        // Remove the used recovery code from the list
+        $updatedCodes = array_values(array_filter($recoveryCodes, function($code) use ($submittedCode) {
+            return $code !== $submittedCode;
```
You should be using `hash_equals()` here to avoid timing attacks that could brute-force a valid recovery code.

tnylea replied:
roger that. Got it updated ;)

valorin commented on resources/views/livewire/auth/login.blade.php:
```diff
@@ -29,20 +30,55 @@ public function login(): void
+        // Get the user model from auth config in a single line
+        $userAttemptingLogin = app(config('auth.providers.users.model'))::where('email', $this->email)->first();
+
+        if (! $userAttemptingLogin || ! \Illuminate\Support\Facades\Hash::check($this->password, $userAttemptingLogin->password)) {
```
This skips the Timebox for timing attack enumeration protection, making it trivial to identify is a user account exists or not in the application.

tnylea replied:
Can you elaborate on this more. Is there something that needs to be updated for security reasons? Here, we want to check the user's password, rather than running `Auth::attempt`, because `Auth::attempt` will automatically log the user in, and we don't want to do that yet if two-factor authentication is enabled.

Let me know. Thanks 😊

valorin commented on resources/views/livewire/auth/two-factor-challenge.blade.php:
```diff
@@ -0,0 +1,124 @@
+        }
+        
+        // Verify the authentication code
+        $valid = app(VerifyTwoFactorCode::class)(decrypt($user->two_factor_secret), $code);
```
There is no rate limiting here, which would allow an attacker to brute-force the 2FA code.

tnylea replied:
Added rate limiting in the latest commit. Thanks!

valorin commented on resources/views/livewire/auth/two-factor-challenge.blade.php:
```diff
@@ -0,0 +1,124 @@
+        }
+        
+        // Process the recovery code
+        $updatedCodes = app(ProcessRecoveryCode::class)(json_decode(decrypt($user->two_factor_recovery_codes), true), $this->recovery_code);
```
There is no rate limiting here, which could allow an attacker to brute-force a recovery code.

On it's own, this is unlikely, however the `!==` instead of `hash_equals()` opens a possibility of a timing attack deducing a valid code.

tnylea replied:
I've added rate limiting here as well 👏

valorin commented on app/Actions/TwoFactorAuth/CompleteTwoFactorAuthentication.php:
```diff
@@ -0,0 +1,24 @@
+    public function __invoke($user): void
+    {
+        // Log the user in
+        Auth::login($user);
```
Correct me if I'm wrong, but I'm pretty sure using `Auth::login()` here will skip all of the extra steps `Auth::attempt()` would normally run during a successful login:

1. Login events aren't figured (i.e. `Attempting`, `Validated`, `Failed`, etc)
2. Remember Me is ignored
3. Legacy password hashes won't be rehashed

duncanmcclean replied:
I believe `Auth::login()` accepts `$remember` as the second parameter, so that one might not be an issue.

I've just checked how Fortify/Jetstream handled this, and it's effectively using `Auth::login()` here as well:

https://github.com/laravel/fortify/blob/f1ad2dd08646fb7d78e609f38e2b1071c71c4e94/src/Http/Controllers/TwoFactorAuthenticatedSessionController.php#L73

duncanmcclean replied:
Also, it looks like [`Auth::login()`](https://github.com/laravel/framework/blob/b6a2af2b96b1088228aa496e376b3c19ba987052/src/Illuminate/Auth/SessionGuard.php#L525) accepts a `$remember` parameter, so that one might not be an issue.

tnylea replied:
Thanks @valorin, at this point we've already verified the user email and password so we can use the `Auth::login()`.

@duncanmcclean Thanks! I've made an update to store the `login.remember` in session so that way this can be passed to the `Auth::login`.

valorin replied:
I'm not saying you need to use `Auth::attempt()` here - it makes sense to use `Auth::login()` given the situation. However running `Auth::login()` on it's own isn't enough. 

During a successful login `Auth::attempt()` fires the `Attempting` event, and triggers password rehashing. Both of these are missing and will not occur when a user has 2FA enabled. The `Attempting` event is an important event for security logging/monitoring, as it gives visibility on login attempts, and rehashing is important to ensure user password hashes are kept updated.

Also, I'm pretty sure the new logic will also fail to trigger `Failed` when a login is failed, which is also an incredibly important event for security logging/monitoring.

My suggestion is to either replicate those events within the starter kit logic, or create a methods on the Guard that will fire the events and perform rehashing, but also support the 2-step process needed here. I can see it being useful for others, so something in the framework could be useful. Also would avoid duplicating across the different starter kits.

As for Fortify not using `Auth::attempt()`, it's probably wrong too. I haven't audited it's code, so don't know the mechanics. All I can tell you is that this specific PR is missing functionality.

nguyentranchung commented:
<img width="661" alt="Screenshot 2025-04-22 at 8 59 01" src="https://github.com/user-attachments/assets/731dc757-a919-422f-8dbb-41a8734a28ea" />

otp maxlength not working

nguyentranchung commented:
Please format the Blade code with whitespace everywhere!

<img width="1138" alt="Screenshot 2025-04-22 at 9 18 05" src="https://github.com/user-attachments/assets/74a74070-e90c-462c-96fb-100d4bc141b9" />

lokomilo commented on app/Actions/TwoFactorAuth/GenerateQrCodeAndSecretKey.php:
```diff
@@ -0,0 +1,50 @@
+    public string $companyName;
+
+    /**
+     * Generate new recovery codes for the user.
```
```suggestion
     * Generate a QR code image and secret key for the user.
```

lucasjose501 commented:
> @tnylea great work, would love to see it getting in there.
> 
> Btw (might be offtopic a little) but does anyone has any suggestion on what best way to implement this in existing project (based on laravel livewire starter kit) is?

The best way is probably waiting for the PR to merge, so everything is reviewed and ready to use, and then look at the changed files and copy/merge into your project by hand.

joetannenbaum commented on resources/views/components/input-otp.blade.php:
```diff
@@ -0,0 +1,90 @@
+            let focusLastInput = (paste.length <= this.total_digits) ? paste.length : this.total_digits;
+            this.$refs['input' + focusLastInput].focus();
+            if (paste.length >= this.total_digits) {
+                let that = this;
```
See arrow function comment above

joetannenbaum commented on resources/views/livewire/auth/login.blade.php:
```diff
@@ -29,20 +32,56 @@ public function login(): void
 
-        if (! Auth::attempt(['email' => $this->email, 'password' => $this->password], $this->remember)) {
+        // Get the user model from auth config in a single line
+        $userAttemptingLogin = User::where('email', $this->email)->first();
```
What if the user isn't coming from `User` model? Can we tap into the auth user resolver?

joetannenbaum commented on resources/views/livewire/auth/two-factor-challenge.blade.php:
```diff
@@ -0,0 +1,167 @@
+            RateLimiter::clear($this->throttleKey($user));
+            
+            // Redirect to the intended page
+            return $this->redirectIntended(default: route('dashboard', absolute: false), navigate: true);
```
If you're returning here, no need for an `else` statement below

tnylea replied:
So, the `ProcessRecoveryCode::class` will return false if the entered recovery code does not match. If it returns false, we need to display a message to the user, which occurs in the `else` condition.

However, feel free to elaborate on your thinking; perhaps there's something I'm missing.

Thanks!

joetannenbaum commented on tests/Feature/Auth/TwoFactorAuthTest.php:
```diff
@@ -0,0 +1,242 @@
+use App\Actions\TwoFactorAuth\GenerateQrCodeAndSecretKey;
+use App\Actions\TwoFactorAuth\GenerateNewRecoveryCodes;
+use Illuminate\Support\Facades\Session;
+use Illuminate\Support\Facades\Hash;
```
Unused import

valorin commented:
My reply to the discussion about not using `Auth::attempt()` is hidden (at least for me) under multiple collapsed sections, so I'm duplicating it here so it's not missed:

---

 I'm not saying you need to use `Auth::attempt()` here - it makes sense to use `Auth::login()` given the situation. However running `Auth::login()` on it's own isn't enough.
 
During a successful login `Auth::attempt()` fires the `Attempting` event, and triggers password rehashing. Both of these are missing and will not occur when a user has 2FA enabled. The `Attempting` event is an important event for security logging/monitoring, as it gives visibility on login attempts, and rehashing is important to ensure user password hashes are kept updated.
 
Also, I'm pretty sure the new logic will also fail to trigger `Failed` when a login is failed, which is also an incredibly important event for security logging/monitoring.
 
My suggestion is to either replicate those events and rehashing within the starter kit logic, or create a methods on the Guard that will fire the events and perform rehashing, but also support the 2-step process needed here. I can see it being useful for others, so something in the framework could be useful. Also would avoid duplicating across the different starter kits.
 
As for Fortify not using `Auth::attempt()`, it's probably wrong too. I haven't audited it's code, so don't know the mechanics. All I can tell you is that this specific PR is missing functionality.

valorin commented on resources/views/livewire/auth/two-factor-challenge.blade.php:
```diff
@@ -0,0 +1,167 @@
+            return redirect()->route('login');
+        }
+        
+        // Process the recovery code
```
Missing `$this->ensureIsNotRateLimited()` here to kick off the rate limiter.

tnylea replied:
Got it updated. Thanks!

valorin commented on resources/views/livewire/auth/two-factor-challenge.blade.php:
```diff
@@ -0,0 +1,167 @@
+     */
+    protected function throttleKey(User $user): string
+    {
+        return Str::transliterate($user->id . '|2fa|' . request()->ip());
```
I suggest dropping the IP and limiting purely based on User ID. Otherwise, it would be possible for an attacker to rotate their IP to bypass the limit.

iamsandakelum commented:
There's a critical error when attempting a two factor verification, on rate limitation code.

It's the "ensureIsNotRateLimited" function at \resources\views\livewire\auth\two-factor-challenge.blade.php

Fix would be this:

`/**
     * Ensure the two-factor authentication request is not rate limited.
     */
    protected function ensureIsNotRateLimited(): void
    {
        $user = app(GetTwoFactorAuthenticatableUser::class)();

        if (! RateLimiter::tooManyAttempts($this->throttleKey($user), 5)) {
            return;
        }

        event(new Lockout(request()));

        $seconds = RateLimiter::availableIn($this->throttleKey($user));

        throw ValidationException::withMessages([
            'auth_code' => __('auth.throttle', [
                'seconds' => $seconds,
                'minutes' => ceil($seconds / 60),
            ]),
        ]);
    }`
    
    Fixes https://github.com/laravel/livewire-starter-kit/pull/84#pullrequestreview-2791725556

tnylea commented on resources/css/app.css:
```diff
@@ -64,3 +64,14 @@ select:focus[data-flux-control] {
+  margin: 0;
+}
+.input-otp input[type=number] {
+  -moz-appearance: textfield;
```
The rules above should handle hiding the input arrows for non-Mozilla browsers, but I've added the appearance `none` here just to keep things consistent 😊

Here is the article I was following on removing the input arrows for numbered inputs: https://www.geeksforgeeks.org/how-to-disable-arrows-from-number-input/.

tnylea commented on resources/views/components/input-otp.blade.php:
```diff
@@ -0,0 +1,90 @@
+            this.$refs['input' + index].value = evt.key;
+            this.$refs['input' + (index + 1)].focus();
+        } else {
+            if (evt.key === 'Backspace') {
```
Thanks, I cleaned this up 👍

tnylea commented:
Thanks for the review, @joetannenbaum; I've implemented all the recommendations, as well as the security recommendations from @valorin (I appreciate it!).

I've also made a few additional security updates, pointed out in this comment: https://github.com/laravel/vue-starter-kit/pull/120#issuecomment-2825570241, from the Vue PR.

@iamsandakelum, thank you for pointing out the issue above. I've gone ahead and fixed it.

nguyentranchung commented:
@tnylea

It seems like you didn’t apply any kind of code formatting at all. Everywhere I look in your source code, I keep running into random or excessive whitespace. This makes the code harder to read and maintain. Please consider using a code formatter or adhering to a consistent style guide.

nguyentranchung commented:
please run `./vendor/bin/pint`

samlev commented:
> please run `./vendor/bin/pint`

It's run automatically by GitHub actions.

iamsandakelum commented:
At the moment, when you are attempting to enable 2FA, the popup appears immediately. It's all good if the contents of the popup is quickly populated too. Unfortunately it's not as the QR code image and the secret has to be generated. So, it's better if you wait until that happens before opening the popup.

This would prevent the subsequent popup opens issue where the old QR code and secret is shown on popup open and gets replaced by the new data midway. If the user is fast enough, or if the connection is slow he could have just scanned the older QR and the whole process would be for naught.

Best way to achieve this is to remove the popup modal trigger altogether, and dispatching an event on end of the enableTwoFactor() function and then listening to that event on the view to open the modal. This way everything works flawlessly, and Flux UI even take care of the "processing" indicator on the button because it's a wire:click call.

jeffagostinho commented:
Waiting for the merge 🎉

angelej commented:
Why? 😿

julesgraus commented:
Oh wow. It looked like a good promising feature. And some of you guys asked good questions 💪. But @taylorotwell why did you close this request? I know it’s a thing that needs maintenance over time and that maintenance must also be taken care of the laravel team if I am correct. But I do think 2fa is a well implemented feature in many sites that helps improving security. I think it adds more value that for example pest and pint which are essentially wrappers around existing tools that I was not waiting for.

hebbet commented:
React and vue got new PRs


https://github.com/laravel/vue-starter-kit/pull/176
https://github.com/laravel/react-starter-kit/pull/156

On both the original was closed as well. 

So there is still hope

ludo237 commented:
> Oh wow. It looked like a good promising feature. And some of you guys asked good questions 💪. But @taylorotwell why did you close this request? I know it’s a thing that needs maintenance over time and that maintenance must also be taken care of the laravel team if I am correct. But I do think 2fa is a well implemented feature in many sites that helps improving security. I think it adds more value that for example pest and pint which are essentially wrappers around existing tools that I was not waiting for. The person creating this request put a lot of work in this feature and deserves an explanation of some kind.

Because it adds to complexity to maintain as a starter kit. I suggest to create e dedicated community starter kit with 2fa

ipontt commented:
This is very nice, I'll be looking at this PR when I need to add 2FA to my livewire project. Still, something I noticed in this PR is its use of lucide icons. I think a starter kit should just stick to sensible defaults in that matter (just the flux-supported heroicons)
